proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/AvailableCommands.java

```diff
@@ -81,8 +89,30 @@ public RootCommandNode<CommandSource> getRootNode() {
   @Override
   public void decode(ByteBuf buf, Direction direction, ProtocolVersion protocolVersion) {
     int commands = ProtocolUtils.readVarInt(buf);
+
+    // When Minecraft serializes this packet the nodes are written in order of id,
+    // as the RootCommandNode is the first node processed it should also be the first node in the array,
+    // Velocity can read the first node of the array to check the type of the node,
+    // if it is the RootCommandNode then we can manipulate this packet without deserializing every node.
+
+    // In order to support Forge, Velocity cannot deserialize every node as it does not support custom parsers,
```

Comment:
Velocity does, in fact, support custom argument parsers. All you have to do is wrap non-vanilla commands [like we do in CrossStitch](https://github.com/VelocityPowered/CrossStitch/blob/master/src/main/java/com/velocitypowered/crossstitch/mixin/command/CommandTreeSerializationMixin.java). This is probably portable enough to carry over into Forge too.

The fact this "support" completely neuters `PlayerAvailableCommandsEvent` is a non-starter for merging this.

Reply:
So pretty much this section is redundant and can be removed and then a PR can be made to the CrossStitch repository to add Forge support?

Reply:
I believe Forge has its own custom argument types - we'd have to be able to parse them. I would be fine with adding these directly into Velocity.

Reply:
CrossStitch should be the solution for mods with custom argument types, yes.

Reply:
Okay, perfect. Thank you for the clarification I'll take a look at getting this resolved tonight

Reply:
I believe I resolved this with the latest commit